src/query/api/v1/handler/prometheus/native/read.go

```diff
@@ -56,9 +57,28 @@ var (
 
 // PromReadHandler represents a handler for prometheus read endpoint.
 type PromReadHandler struct {
-	engine    *executor.Engine
-	tagOpts   models.TagOptions
-	limitsCfg *config.LimitsConfiguration
+	engine          *executor.Engine
+	tagOpts         models.TagOptions
+	limitsCfg       *config.LimitsConfiguration
+	promReadMetrics promReadMetrics
+}
+
+type promReadMetrics struct {
+	fetchSuccess      tally.Counter
+	fetchErrorsServer tally.Counter
+	fetchErrorsClient tally.Counter
+	fetchTimerSuccess tally.Timer
+	maxDatapoints     tally.Gauge
+}
+
+func newPromReadMetrics(scope tally.Scope) promReadMetrics {
+	return promReadMetrics{
+		fetchSuccess:      scope.Counter("fetch.success"),
+		fetchErrorsServer: scope.Tagged(map[string]string{"code": "5XX"}).Counter("fetch.errors"),
```

Comment:
These are generic HTTP metrics--maybe worth factoring them out and applying them to every endpoint (as middleware)?